File: src/dart/lib/src/models/connection_pool.dart
import 'package:flutter_webrtc/flutter_webrtc.dart' as rtc;

/// A pooled ephemeral probe connection (spec §2.5, §3.3).
class ConnectionPoolEntry {
  final String targetId;
  final rtc.RTCPeerConnection pc;
  final rtc.RTCDataChannel dc;
  DateTime lastUsed;
  final DateTime createdAt;

  ConnectionPoolEntry({
    required this.targetId,
    required this.pc,
    required this.dc,
    DateTime? lastUsed,
    DateTime? createdAt,
  })  : lastUsed = lastUsed ?? DateTime.now(),
        createdAt = createdAt ?? DateTime.now();
}

/// Pool of established probe connections, keyed by target peer. Evicts (and
/// closes) the least-recently-used entry once [maxSize] is exceeded.
class ConnectionPool {
  final List<ConnectionPoolEntry> entries = [];
  final int maxSize;

  ConnectionPool({this.maxSize = 10});

  /// Returns the pooled connection for [targetId], or null.
  ConnectionPoolEntry? find(String targetId) {
    for (final entry in entries) {
      if (entry.targetId == targetId) return entry;
    }
    return null;
  }

  /// Adds [entry], evicting the least-recently-used entry (closing its
  /// PeerConnection) when the pool is at capacity.
  void add(ConnectionPoolEntry entry) {
    if (entries.length >= maxSize) {
      var oldestIndex = 0;
      for (var i = 1; i < entries.length; i++) {
        if (entries[i].lastUsed.isBefore(entries[oldestIndex].lastUsed)) {
          oldestIndex = i;
        }
      }
      final oldest = entries.removeAt(oldestIndex);
      try {
        oldest.pc.close();
      } catch (_) {
        // Already closed.
      }
    }
    entries.add(entry);
  }

  /// Closes and drops entries unused for longer than [maxAge].
  void cleanup({Duration maxAge = const Duration(minutes: 2)}) {
    final now = DateTime.now();
    final stale = entries
        .where((entry) => now.difference(entry.lastUsed) > maxAge)
        .toList();
    entries.removeWhere(stale.contains);
    for (final entry in stale) {
      try {
        entry.pc.close();
      } catch (_) {
        // Already closed.
      }
    }
  }

  /// Closes every pooled connection.
  void dispose() {
    for (final entry in entries) {
      try {
        entry.pc.close();
      } catch (_) {
        // Already closed.
      }
    }
    entries.clear();
  }
}

File: src/js/src/connection-pool.js
/**
 * LRU-ish pool of established probe connections (spec §2.5, §3.3).
 * Entries: { targetId, pc, dc, lastUsed, createdAt }.
 */
export class ConnectionPool {
  constructor(maxSize = 10) {
    this.entries = [];
    this.maxSize = maxSize;
  }

  find(targetId) {
    return this.entries.find((entry) => entry.targetId === targetId) || null;
  }

  add(entry) {
    if (this.entries.length >= this.maxSize) {
      let oldestIndex = 0;
      for (let i = 1; i < this.entries.length; i++) {
        if (this.entries[i].lastUsed < this.entries[oldestIndex].lastUsed) {
          oldestIndex = i;
        }
      }
      const [oldest] = this.entries.splice(oldestIndex, 1);
      try {
        oldest.pc.close();
      } catch {
        // Already closed.
      }
    }
    this.entries.push(entry);
  }

  cleanup(maxAgeMs = 120000) {
    const now = Date.now();
    this.entries = this.entries.filter((entry) => {
      if (now - entry.lastUsed > maxAgeMs) {
        try {
          entry.pc.close();
        } catch {
          // Already closed.
        }
        return false;
      }
      return true;
    });
  }

  close() {
    for (const entry of this.entries) {
      try {
        entry.pc.close();
      } catch {
        // Already closed.
      }
    }
    this.entries = [];
  }
}